File: C++_syntax/virtual_method.py
#-*-coding:utf-8 -*-

class Animal:

    def eat(self):
        print("I eat like a generic animal.")

class Fish(Animal):

    def eat(self):
        print("I eat like a fish.")

class Wolf(Animal):

    def eat(self):
        print("I eat like a wolf.")

class GoldFish(Fish):

    def eat(self):
        print("I eat like a goldfish.")

class OtherAnimal(Animal):
    pass

if __name__ == "__main__":
    animal = Animal()
    animal.eat()
    fish = Fish()
    fish.eat()
    wolf = Wolf()
    wolf.eat()
    gfish = GoldFish()
    gfish.eat()
    other_animal = OtherAnimal()
    other_animal.eat()

File: C++_syntax/virtual_method.cc
#include<iostream>
#include<vector>

using namespace std;

class Animal {
    public:
        void eat() const {
            cout<<"I eat like a generic Animal.\n";
        }
        ~Animal() {}
};

class Wolf: public Animal {
    public:
        void eat() const {
            cout<<"I eat like a wolf!\n";
        }
};

class Fish: public Animal {
    public:
        void eat() const {
            cout<<"I eat like a fish!\n";
        }
};

class GoldFish: public Fish {
    public:
        void eat() const {
            cout<<"I eat like a goldfish!\n";
        }
};

class OtherAnimal: public Animal {
};

int main() {
    vector<Animal*> animals;
    animals.push_back(new Animal());
    animals.push_back(new Wolf());
    animals.push_back(new Fish());
    animals.push_back(new GoldFish());
    animals.push_back(new OtherAnimal());

    for (vector<Animal*>::const_iterator it = animals.begin(); it != animals.end(); ++it) {
        (*it)->eat();
        delete *it;
    }

    return 0;
}

